File: test-mcp-server/src/tools.ts
/**
 * Tool implementations
 * Each function corresponds to an MCP tool
 */

import axios from 'axios';

const BASE_URL = process.env.API_BASE_URL || "https://petstore.swagger.io/v2";

const API_KEY = process.env.API_KEY || "";

function getAuthHeaders(): Record<string, string> {
  return API_KEY ? { Authorization: `Bearer ${API_KEY}` } : {};
}


/**
 * uploads an image
 * POST /pet/{petId}/uploadImage
 */
export async function uploadFile(args: Record<string, unknown>): Promise<unknown> {
  const response = await axios({
    method: "POST",
    url: `${BASE_URL}/pet/${args.petId ?? ''}/uploadImage`,
    
    
    headers: getAuthHeaders(),
  });
  return response.data;
}

/**
 * Add a new pet to the store
 * POST /pet
 */
export async function addPet(args: Record<string, unknown>): Promise<unknown> {
  const response = await axios({
    method: "POST",
    url: `${BASE_URL}/pet`,
    
    data: args.body,
    headers: getAuthHeaders(),
  });
  return response.data;
}

/**
 * Update an existing pet
 * PUT /pet
 */
export async function updatePet(args: Record<string, unknown>): Promise<unknown> {
  const response = await axios({
    method: "PUT",
    url: `${BASE_URL}/pet`,
    
    data: args.body,
    headers: getAuthHeaders(),
  });
  return response.data;
}

/**
 * Finds Pets by status
 * GET /pet/findByStatus
 */
export async function findPetsByStatus(args: Record<string, unknown>): Promise<unknown> {
  const response = await axios({
    method: "GET",
    url: `${BASE_URL}/pet/findByStatus`,
    params: { status: args.status },
    
    headers: getAuthHeaders(),
  });
  return response.data;
}

/**
 * Finds Pets by tags
 * GET /pet/findByTags
 */
export async function findPetsByTags(args: Record<string, unknown>): Promise<unknown> {
  const response = await axios({
    method: "GET",
    url: `${BASE_URL}/pet/findByTags`,
    params: { tags: args.tags },
    
    headers: getAuthHeaders(),
  });
  return response.data;
}

/**
 * Find pet by ID
 * GET /pet/{petId}
 */
export async function getPetById(args: Record<string, unknown>): Promise<unknown> {
  const response = await axios({
    method: "GET",
    url: `${BASE_URL}/pet/${args.petId ?? ''}`,
    
    
    headers: getAuthHeaders(),
  });
  return response.data;
}

/**
 * Updates a pet in the store with form data
 * POST /pet/{petId}
 */
export async function updatePetWithForm(args: Record<string, unknown>): Promise<unknown> {
  const response = await axios({
    method: "POST",
    url: `${BASE_URL}/pet/${args.petId ?? ''}`,
    
    
    headers: getAuthHeaders(),
  });
  return response.data;
}

/**
 * Deletes a pet
 * DELETE /pet/{petId}
 */
export async function deletePet(args: Record<string, unknown>): Promise<unknown> {
  const response = await axios({
    method: "DELETE",
    url: `${BASE_URL}/pet/${args.petId ?? ''}`,
    
    
    headers: getAuthHeaders(),
  });
  return response.data;
}

/**
 * Returns pet inventories by status
 * GET /store/inventory
 */
export async function getInventory(args: Record<string, unknown>): Promise<unknown> {
  const response = await axios({
    method: "GET",
    url: `${BASE_URL}/store/inventory`,
    
    
    headers: getAuthHeaders(),
  });
  return response.data;
}

/**
 * Place an order for a pet
 * POST /store/order
 */
export async function placeOrder(args: Record<string, unknown>): Promise<unknown> {
  const response = await axios({
    method: "POST",
    url: `${BASE_URL}/store/order`,
    
    data: args.body,
    headers: getAuthHeaders(),
  });
  return response.data;
}

/**
 * Find purchase order by ID
 * GET /store/order/{orderId}
 */
export async function getOrderById(args: Record<string, unknown>): Promise<unknown> {
  const response = await axios({
    method: "GET",
    url: `${BASE_URL}/store/order/${args.orderId ?? ''}`,
    
    
    headers: getAuthHeaders(),
  });
  return response.data;
}

/**
 * Delete purchase order by ID
 * DELETE /store/order/{orderId}
 */
export async function deleteOrder(args: Record<string, unknown>): Promise<unknown> {
  const response = await axios({
    method: "DELETE",
    url: `${BASE_URL}/store/order/${args.orderId ?? ''}`,
    
    
    headers: getAuthHeaders(),
  });
  return response.data;
}

/**
 * Creates list of users with given input array
 * POST /user/createWithList
 */
export async function createUsersWithListInput(args: Record<string, unknown>): Promise<unknown> {
  const response = await axios({
    method: "POST",
    url: `${BASE_URL}/user/createWithList`,
    
    data: args.body,
    headers: getAuthHeaders(),
  });
  return response.data;
}

/**
 * Get user by user name
 * GET /user/{username}
 */
export async function getUserByName(args: Record<string, unknown>): Promise<unknown> {
  const response = await axios({
    method: "GET",
    url: `${BASE_URL}/user/${args.username ?? ''}`,
    
    
    headers: getAuthHeaders(),
  });
  return response.data;
}

/**
 * Updated user
 * PUT /user/{username}
 */
export async function updateUser(args: Record<string, unknown>): Promise<unknown> {
  const response = await axios({
    method: "PUT",
    url: `${BASE_URL}/user/${args.username ?? ''}`,
    
    data: args.body,
    headers: getAuthHeaders(),
  });
  return response.data;
}

/**
 * Delete user
 * DELETE /user/{username}
 */
export async function deleteUser(args: Record<string, unknown>): Promise<unknown> {
  const response = await axios({
    method: "DELETE",
    url: `${BASE_URL}/user/${args.username ?? ''}`,
    
    
    headers: getAuthHeaders(),
  });
  return response.data;
}

/**
 * Logs user into the system
 * GET /user/login
 */
export async function loginUser(args: Record<string, unknown>): Promise<unknown> {
  const response = await axios({
    method: "GET",
    url: `${BASE_URL}/user/login`,
    params: { username: args.username, password: args.password },
    
    headers: getAuthHeaders(),
  });
  return response.data;
}

/**
 * Logs out current logged in user session
 * GET /user/logout
 */
export async function logoutUser(args: Record<string, unknown>): Promise<unknown> {
  const response = await axios({
    method: "GET",
    url: `${BASE_URL}/user/logout`,
    
    
    headers: getAuthHeaders(),
  });
  return response.data;
}

/**
 * Creates list of users with given input array
 * POST /user/createWithArray
 */
export async function createUsersWithArrayInput(args: Record<string, unknown>): Promise<unknown> {
  const response = await axios({
    method: "POST",
    url: `${BASE_URL}/user/createWithArray`,
    
    data: args.body,
    headers: getAuthHeaders(),
  });
  return response.data;
}

/**
 * Create user
 * POST /user
 */
export async function createUser(args: Record<string, unknown>): Promise<unknown> {
  const response = await axios({
    method: "POST",
    url: `${BASE_URL}/user`,
    
    data: args.body,
    headers: getAuthHeaders(),
  });
  return response.data;
}

File: test-mcp-python/src/tools.py
"""
Tool implementations for MCP server.

Each function corresponds to an API endpoint.
"""

import os
import httpx
from typing import Any

from .models import *

# Configuration
BASE_URL = os.environ.get("API_BASE_URL", "https://petstore.swagger.io/v2")

API_KEY = os.environ.get("API_KEY", "")

def get_auth_headers() -> dict[str, str]:
    return {"Authorization": f"Bearer {API_KEY}"} if API_KEY else {}

# Async HTTP client
client = httpx.AsyncClient(
    base_url=BASE_URL,
    headers=get_auth_headers(),
    timeout=30.0,
)


async def uploadFile(args: UploadFileInput) -> dict[str, Any]:
    """uploads an image"""
    url = f"/pet/{args.petId}/uploadImage"
    response = await client.post(url)
    response.raise_for_status()
    return response.json()


async def addPet(args: AddPetInput) -> dict[str, Any]:
    """Add a new pet to the store"""
    json_data = args.body.model_dump() if hasattr(args.body, "model_dump") else args.body
    url = "/pet"
    response = await client.post(url, json=json_data)
    response.raise_for_status()
    return response.json()


async def updatePet(args: UpdatePetInput) -> dict[str, Any]:
    """Update an existing pet"""
    json_data = args.body.model_dump() if hasattr(args.body, "model_dump") else args.body
    url = "/pet"
    response = await client.put(url, json=json_data)
    response.raise_for_status()
    return response.json()


async def findPetsByStatus(args: FindPetsByStatusInput) -> dict[str, Any]:
    """Finds Pets by status"""
    params = {"status": args.status}
    url = "/pet/findByStatus"
    response = await client.get(url, params=params)
    response.raise_for_status()
    return response.json()


async def findPetsByTags(args: FindPetsByTagsInput) -> dict[str, Any]:
    """Finds Pets by tags"""
    params = {"tags": args.tags}
    url = "/pet/findByTags"
    response = await client.get(url, params=params)
    response.raise_for_status()
    return response.json()


async def getPetById(args: GetPetByIdInput) -> dict[str, Any]:
    """Find pet by ID"""
    url = f"/pet/{args.petId}"
    response = await client.get(url)
    response.raise_for_status()
    return response.json()


async def updatePetWithForm(args: UpdatePetWithFormInput) -> dict[str, Any]:
    """Updates a pet in the store with form data"""
    url = f"/pet/{args.petId}"
    response = await client.post(url)
    response.raise_for_status()
    return response.json()


async def deletePet(args: DeletePetInput) -> dict[str, Any]:
    """Deletes a pet"""
    url = f"/pet/{args.petId}"
    response = await client.delete(url)
    response.raise_for_status()
    return response.json()


async def getInventory(args: GetInventoryInput) -> dict[str, Any]:
    """Returns pet inventories by status"""
    url = "/store/inventory"
    response = await client.get(url)
    response.raise_for_status()
    return response.json()


async def placeOrder(args: PlaceOrderInput) -> dict[str, Any]:
    """Place an order for a pet"""
    json_data = args.body.model_dump() if hasattr(args.body, "model_dump") else args.body
    url = "/store/order"
    response = await client.post(url, json=json_data)
    response.raise_for_status()
    return response.json()


async def getOrderById(args: GetOrderByIdInput) -> dict[str, Any]:
    """Find purchase order by ID"""
    url = f"/store/order/{args.orderId}"
    response = await client.get(url)
    response.raise_for_status()
    return response.json()


async def deleteOrder(args: DeleteOrderInput) -> dict[str, Any]:
    """Delete purchase order by ID"""
    url = f"/store/order/{args.orderId}"
    response = await client.delete(url)
    response.raise_for_status()
    return response.json()


async def createUsersWithListInput(args: CreateUsersWithListInputInput) -> dict[str, Any]:
    """Creates list of users with given input array"""
    json_data = args.body.model_dump() if hasattr(args.body, "model_dump") else args.body
    url = "/user/createWithList"
    response = await client.post(url, json=json_data)
    response.raise_for_status()
    return response.json()


async def getUserByName(args: GetUserByNameInput) -> dict[str, Any]:
    """Get user by user name"""
    url = f"/user/{args.username}"
    response = await client.get(url)
    response.raise_for_status()
    return response.json()


async def updateUser(args: UpdateUserInput) -> dict[str, Any]:
    """Updated user"""
    json_data = args.body.model_dump() if hasattr(args.body, "model_dump") else args.body
    url = f"/user/{args.username}"
    response = await client.put(url, json=json_data)
    response.raise_for_status()
    return response.json()


async def deleteUser(args: DeleteUserInput) -> dict[str, Any]:
    """Delete user"""
    url = f"/user/{args.username}"
    response = await client.delete(url)
    response.raise_for_status()
    return response.json()


async def loginUser(args: LoginUserInput) -> dict[str, Any]:
    """Logs user into the system"""
    params = {"username": args.username, "password": args.password}
    url = "/user/login"
    response = await client.get(url, params=params)
    response.raise_for_status()
    return response.json()


async def logoutUser(args: LogoutUserInput) -> dict[str, Any]:
    """Logs out current logged in user session"""
    url = "/user/logout"
    response = await client.get(url)
    response.raise_for_status()
    return response.json()


async def createUsersWithArrayInput(args: CreateUsersWithArrayInputInput) -> dict[str, Any]:
    """Creates list of users with given input array"""
    json_data = args.body.model_dump() if hasattr(args.body, "model_dump") else args.body
    url = "/user/createWithArray"
    response = await client.post(url, json=json_data)
    response.raise_for_status()
    return response.json()


async def createUser(args: CreateUserInput) -> dict[str, Any]:
    """Create user"""
    json_data = args.body.model_dump() if hasattr(args.body, "model_dump") else args.body
    url = "/user"
    response = await client.post(url, json=json_data)
    response.raise_for_status()
    return response.json()
